File: metrics.py
import subprocess
import os
import time
import re
from CICD.registry import load_registry

def get_service_status(service_name: str) -> str:
    result = subprocess.run(
        ["systemctl", "is-active", service_name],
        capture_output=True, text=True
    )
    return result.stdout.strip()

def get_service_uptime(service_name: str) -> str:
    result = subprocess.run(
        ["systemctl", "show", service_name, "--property=ActiveEnterTimestamp"],
        capture_output=True, text=True
    )
    line = result.stdout.strip()
    if "=" in line:
        return line.split("=", 1)[1]
    return "unknown"

def get_service_memory(service_name: str) -> str:
    result = subprocess.run(
        ["systemctl", "show", service_name, "--property=MemoryCurrent"],
        capture_output=True, text=True
    )
    line = result.stdout.strip()
    if "=" in line:
        val = line.split("=")[1].strip()
        if val.isdigit():
            mb = int(val) / (1024 * 1024)
            return f"{mb:.1f}MB"
    return "unknown"

def get_system_stats() -> dict:
    # CPU
    cpu = subprocess.run(
        ["top", "-bn1"],
        capture_output=True, text=True
    ).stdout
    cpu_line = [l for l in cpu.split("\n") if "%Cpu" in l or "Cpu(s)" in l]
    if cpu_line:
        line = cpu_line[0]
        
    # extract idle and calculate usage
    idle = re.search(r'(\d+\.?\d*)\s*id', line)
    if idle:
        usage = 100 - float(idle.group(1))
        cpu_usage = f"{usage:.1f}%"

    # RAM
    mem = subprocess.run(["free", "-h"], capture_output=True, text=True).stdout
    mem_line = [l for l in mem.split("\n") if l.startswith("Mem:")][0].split()
    ram_total = mem_line[1]
    ram_used = mem_line[2]

    # Disk
    disk = subprocess.run(["df", "-h", "/"], capture_output=True, text=True).stdout
    disk_line = disk.split("\n")[1].split()
    disk_used = disk_line[2]
    disk_total = disk_line[1]

    return {
        "cpu": cpu_usage,
        "ram": f"{ram_used}/{ram_total}",
        "disk": f"{disk_used}/{disk_total}"
    }

def clear():
    os.system("clear")

def render(registry: dict, stats: dict):
    clear()
    print("\n[→] EZDeploy Metrics\n")
    print(f"  {'PROJECT':<20} {'STATUS':<10} {'PORT':<8} {'MEMORY':<12} {'STARTED'}")
    print(f"  {'-------':<20} {'------':<10} {'----':<8} {'------':<12} {'-------'}")

    for name, info in registry.items():
        service_name = f"{name}.service"
        status = get_service_status(service_name)
        memory = get_service_memory(service_name)
        uptime = get_service_uptime(service_name)
        port = info.get("port", "?")
        status_display = f"[✓] {status}" if status == "active" else f"[✗] {status}"
        print(f"  {name:<20} {status_display:<10} {str(port):<8} {memory:<12} {uptime}")

    print(f"\n  CPU:  {stats['cpu']}")
    print(f"  RAM:  {stats['ram']}")
    print(f"  Disk: {stats['disk']}")
    print(f"\n  refreshing every 5s — ctrl+c to exit\n")

def metrics():
    print("[→] Starting EZDeploy metrics...")
    try:
        while True:
            registry = load_registry()
            stats = get_system_stats()
            render(registry, stats)
            time.sleep(5)
    except KeyboardInterrupt:
        clear()
        print("[✓] Metrics stopped\n")

if __name__ == "__main__":
    metrics()

File: services/metrics.go
package services

import (
	"EZDeploy/core"
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"
)

func Metrics() error {
	fmt.Println("[→] Starting EZDeploy metrics...")

	for {
		registry, err := core.GetRegistry()
		if err != nil {
			return err
		}

		stats := systemStats{CPU: getCPUUsage(), RAM: getRAMUsage(), Disk: getDiskUsage()}
		renderMetrics(registry, stats)

		time.Sleep(5 * time.Second)
	}
}

type systemStats struct {
	CPU  string
	RAM  string
	Disk string
}

func renderMetrics(registry core.Registry, stats systemStats) {
	clear()

	var b strings.Builder

	fmt.Fprintln(&b)
	fmt.Fprintln(&b, "[→] EZDeploy Metrics")
	fmt.Fprintln(&b)

	fmt.Fprintf(&b, "  %-20s %-12s %-8s %-12s %s\n", "PROJECT", "STATUS", "PORT", "MEMORY", "STARTED")
	fmt.Fprintf(&b, "  %-20s %-12s %-8s %-12s %s\n", "-------", "------", "----", "------", "-------")

	for name, info := range registry {
		serviceName := info.ServiceName
		if serviceName == "" {
			serviceName = core.ManagedName(name)
		}

		status := getServiceStatus(serviceName)
		memory := getServiceMemory(serviceName)
		started := getServiceUptime(serviceName)

		statusDisplay := "[✗] " + status
		if status == "active" {
			statusDisplay = "[✓] " + status
		}

		port := "?"
		if info.Port != 0 {
			port = strconv.Itoa(info.Port)
		}

		fmt.Fprintf(&b, "  %-20s %-12s %-8s %-12s %s\n", name, statusDisplay, port, memory, started)
	}

	fmt.Fprintf(&b, "\n  CPU:  %s\n", stats.CPU)
	fmt.Fprintf(&b, "  RAM:  %s\n", stats.RAM)
	fmt.Fprintf(&b, "  Disk: %s\n", stats.Disk)
	fmt.Fprintln(&b)
	fmt.Fprintln(&b, "  refreshing every 5s — ctrl+c to exit")
	fmt.Fprintln(&b)

	terminalPrint(b.String())
}

func getServiceStatus(serviceName string) string {
	out, err := commandOutput("systemctl", "is-active", serviceName)
	if err != nil {
		return "unknown"
	}

	status := strings.TrimSpace(out)
	if status == "" {
		return "unknown"
	}

	return status
}

func getServiceUptime(serviceName string) string {
	value, ok := systemdProperty(serviceName, "ActiveEnterTimestamp")
	if !ok {
		return "unknown"
	}
	return value
}

func getServiceMemory(serviceName string) string {
	value, ok := systemdProperty(serviceName, "MemoryCurrent")
	if !ok {
		return "unknown"
	}
	bytes, err := strconv.ParseFloat(value, 64)
	if err != nil {
		return "unknown"
	}

	mb := bytes / 1024 / 1024
	return fmt.Sprintf("%.1fMB", mb)
}

func systemdProperty(serviceName, property string) (string, bool) {
	out, err := commandOutput("systemctl", "show", serviceName, "--property="+property)
	if err != nil {
		return "", false
	}
	_, value, ok := strings.Cut(strings.TrimSpace(out), "=")
	value = strings.TrimSpace(value)
	return value, ok && value != ""
}

func getCPUUsage() string {
	out, err := commandOutput("top", "-bn1")
	if err != nil {
		return "unknown"
	}

	lines := strings.Split(out, "\n")
	for _, line := range lines {
		if strings.Contains(line, "%Cpu") || strings.Contains(line, "Cpu(s)") {
			re := regexp.MustCompile(`(\d+\.?\d*)\s*id`)
			match := re.FindStringSubmatch(line)
			if len(match) < 2 {
				return "unknown"
			}

			idle, err := strconv.ParseFloat(match[1], 64)
			if err != nil {
				return "unknown"
			}

			usage := 100 - idle
			return fmt.Sprintf("%.1f%%", usage)
		}
	}

	return "unknown"
}

func getRAMUsage() string {
	out, err := commandOutput("free", "-h")
	if err != nil {
		return "unknown"
	}

	for _, line := range strings.Split(out, "\n") {
		if strings.HasPrefix(line, "Mem:") {
			parts := strings.Fields(line)
			if len(parts) >= 3 {
				return fmt.Sprintf("%s/%s", parts[2], parts[1])
			}
		}
	}

	return "unknown"
}

func getDiskUsage() string {
	out, err := commandOutput("df", "-h", "/")
	if err != nil {
		return "unknown"
	}

	lines := strings.Split(strings.TrimSpace(out), "\n")
	if len(lines) < 2 {
		return "unknown"
	}

	parts := strings.Fields(lines[1])
	if len(parts) < 3 {
		return "unknown"
	}

	used := parts[2]
	total := parts[1]

	return fmt.Sprintf("%s/%s", used, total)
}

func commandOutput(name string, args ...string) (string, error) {
	cmd := exec.Command(name, args...)
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}

	return string(out), nil
}

func clear() {
	fmt.Print("\033[H\033[2J\r")
	_ = os.Stdout.Sync()
}

func terminalPrint(s string) {
	// Normalize existing CRLF first so we do not create \r\r\n.
	s = strings.ReplaceAll(s, "\r\n", "\n")
	s = strings.ReplaceAll(s, "\n", "\r\n")

	fmt.Print(s)
}
